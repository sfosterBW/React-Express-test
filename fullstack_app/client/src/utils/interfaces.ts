//Objects
export interface IFruit {
  id: string
  name: string
  best: boolean
}

export type NewFruit = Omit<IFruit, 'id'>

export interface IFruitResponse {
  status: number
}

//States
export interface AlertState {
  toggle: boolean
}

export interface FruitState {
  data: IFruit[]
}

export interface ModalState {
  toggle: boolean,
  fruit?: IFruit
}
