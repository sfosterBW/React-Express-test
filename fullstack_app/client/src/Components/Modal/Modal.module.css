.wrapper {
  position: absolute;
  bottom: 0;
  right: 0;
  left: 0;
  top: 0;
  display: flex;
  justify-content: center;
  align-items: center;
}

.modal {
  position: relative;
  border-radius: 5px;
  background: lightgray;
}

.modal form {
  background: lightgray;
}

.title {
  font-size: 3vmin;
  text-align: center;
}

.close {
  position: absolute;
  top: 1.5vmin;
  right: 0px;
  transform: translateY(-50%);
  cursor: pointer;
}
