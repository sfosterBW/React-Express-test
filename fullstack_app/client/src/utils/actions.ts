import { IFruit } from './interfaces'
 
//Action const
export const TOGGLE_ALERT = "TOGGLE_ALERT"

//Action shape
interface ToggleAlert {
  type: typeof TOGGLE_ALERT,
  payload: boolean
}

//Action type
export type AlertActionTypes = ToggleAlert

//Action
export function toggleAlert(alertValue: boolean): AlertActionTypes {
  return {
    type: TOGGLE_ALERT,
    payload: alertValue
  }
}

//Action const
export const TOGGLE_MODAL = "TOGGLE_MODAL"
export const OPEN_MODAL = "OPEN_MODAL"

//Action shape
interface ToggleModal {
  type: typeof TOGGLE_MODAL,
  payload: boolean
}

interface OpenModal {
  type: typeof OPEN_MODAL,
  payload: IFruit
}

//Action type
export type ModalActionTypes = ToggleModal | OpenModal

//Action
export function toggleModal(modalValue: boolean): ModalActionTypes {
  return {
    type: TOGGLE_MODAL,
    payload: modalValue,
  }
}

export function openModal(fruit: IFruit): ModalActionTypes {
  return {
    type: OPEN_MODAL,
    payload: fruit,
  }
}

//Action const
export const GET_FRUITS = 'GET_FRUITS'
export const CREATE_FRUIT = 'CREATE_FRUIT'
export const REMOVE_FRUIT = 'REMOVE_FRUIT'
export const UPDATE_FRUIT = 'UPDATE_FRUIT'

//Action shape
interface GetFruits {
  type: typeof GET_FRUITS,
  payload: IFruit[]
}

interface CreateFruit {
  type: typeof CREATE_FRUIT,
  payload: IFruit
}

interface RemoveFruit {
  type: typeof REMOVE_FRUIT,
  payload: number
}

interface UpdateFruit {
  type: typeof UPDATE_FRUIT,
  payload: IFruit
}

//Action types
export type FruitActionTypes = GetFruits | CreateFruit | RemoveFruit | UpdateFruit

//Actions
export const getFruits = (fruits: IFruit[]): FruitActionTypes => {
  return {
    type: GET_FRUITS,
    payload: fruits
  }
}

export const createFruit = (fruit: IFruit): FruitActionTypes => {
  return {
    type: CREATE_FRUIT,
    payload: fruit,
  }
}

export const removeFruit = (id: number): FruitActionTypes => {
  return {
    type: REMOVE_FRUIT,
    payload: id,
  }
}

export const updateFruit = (fruit: IFruit): FruitActionTypes => {
  return {
    type: UPDATE_FRUIT,
    payload: fruit,
  }
}
