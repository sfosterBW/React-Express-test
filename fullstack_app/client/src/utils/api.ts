import axios from 'axios'
import { IFruit, NewFruit } from './interfaces'

axios.defaults.baseURL = '/fruit-api'

const fetchFruit = async (): Promise<IFruit[]> => {
  const response = await axios.get('/')
  return response.data
}

const createFruit = async (fruit: NewFruit): Promise<IFruit> => {
    const response = await axios.post('/', fruit)
    return response.data
}

const updateFruit = async (fruit: IFruit): Promise<IFruit> => {
    const response = await axios.put(`/${fruit.id}`, fruit)
    return response.data
}

const deleteFruit = async (id: string): Promise<IFruit['id']> => {
  const response = await axios.delete(`/${id}`)
  return response.data
}

export default { fetchFruit, createFruit, updateFruit, deleteFruit }
