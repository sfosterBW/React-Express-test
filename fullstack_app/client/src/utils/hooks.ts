import { useState } from 'react'

export const useField = (
  initValue: string,
  label: string,
  name: string,
  type: string) => {

  const [value, setValue] = useState(initValue)

  const handleChange = (event: any) => {
    setValue(event.target.value)
  }

  const reset = () => {
    setValue("")
  }

  return {
    handleChange,
    label,
    name,
    reset,
    type,
    value
  }
}
