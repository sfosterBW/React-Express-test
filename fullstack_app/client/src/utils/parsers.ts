import { Fruit } from './interfaces'

const isBoolean = (bool: any): bool is boolean => {
  return typeof bool === 'boolean'
}

const isString = (text: any): text is string => {
  return typeof text === 'string' || text instanceof String;
}

const parseBest = (best: any): boolean => {
  if (typeof best === 'undefined' || !isBoolean(best)) {
    throw new Error(`Incorrect or missing best: ${String(best)}`);
  }

  return best
}

export const parseId = (id: any): string => {
  if (!id || !isString(id)) {
    throw new Error(`Incorrect or missing id: ${String(id)}`)
  }

  return id
}

const parseName = (name: any): string => {
  if (!name || !isString(name)) {
    throw new Error(`Incorrect or missing name: ${String(name)}`)
  }

  return name
}

const parseArray = (array: any): any[] => {
  if(array && !Array.isArray(array)) {
    throw new Error(`Incorrect or missing an array ${array}`);
  }

  return array
}

export const toFruit = (object: any): Fruit => {
  return {
    id: parseId(object.id),
    name: parseName(object.name),
    best: parseBest(object.best)
  }
}

export const toFruits = (object: any): Fruit[] => {
  const array = parseArray(object)
  return array.map(fruit => toFruit(fruit))
}

export const toId = (object: any): string => {
  return parseId(object)
}

export default { toFruit, toFruits, toId }
