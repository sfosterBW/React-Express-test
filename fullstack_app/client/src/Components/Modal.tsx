import React, { FC } from 'react'
import { RootState } from '../utils/store'
import { useSelector, useDispatch } from 'react-redux'
import { toggleModal } from '../utils/actions'
import Form from './Form'
import { IFruit } from '../utils/interfaces'

interface Props {
  fruit: IFruit | undefined
  title: string
}

const Modal: FC<Props> = ({ fruit, title }) => {

  const selectToggle = (state: RootState) => state.modal.toggle
  const toggle = useSelector(selectToggle)
  const dispatch = useDispatch()

  const displayModal = () => {
    if (toggle && fruit) {
      return (
        <div className="modal-wrapper">
          <div className="modal">
            <h2 className="modal-title">{title}</h2>
            <Form fruit={fruit} />
            <button
              className="closeButton"
              name="close"
              onClick={() => dispatch(toggleModal(!toggle))}>
              &times;
              </button>
          </div>
        </div>
      )
    } else {
      console.log("Modal fruit undefined")
      return null
    }
  }

  return (
    <div>{displayModal()}</div>
  )
}

export default Modal
