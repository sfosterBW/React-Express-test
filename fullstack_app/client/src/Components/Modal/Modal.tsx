import React, { FC } from 'react'
import { RootState } from '../../utils/store'
import { useSelector, useDispatch } from 'react-redux'
import { toggleModal } from '../../utils/actions'
import { IFruit } from '../../utils/interfaces'
import Form from '../Form/Form'
import styles from './Modal.module.css'

interface Props {
  fruit: IFruit | undefined
  title: string
}

const Modal: FC<Props> = ({ fruit, title }) => {

  const selectToggle = (state: RootState) => state.modal.toggle
  const toggle = useSelector(selectToggle)
  const dispatch = useDispatch()

  const displayModal = () => {
    if (toggle && fruit) {
      return (
        <section className={styles.wrapper}>
          <div className={styles.modal}>
            <h2 className={styles.title}>{title}</h2>
            <Form fruit={fruit} />
            <button
              className={styles.close}
              name="close"
              onClick={() => dispatch(toggleModal(!toggle))}>
              &times;
              </button>
          </div>
        </section>
      )
    }
  }

  return (
    <div>{displayModal()}</div>
  )
}

export default Modal
