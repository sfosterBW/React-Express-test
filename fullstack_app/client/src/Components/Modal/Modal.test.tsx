import React from 'react'
import { render, fireEvent, cleanup } from '@testing-library/react'
import { fruit } from '../../utils/test-helper'

import Modal from './Modal'

afterEach(cleanup)

const mockDispatch = jest.fn()
jest.mock('react-redux', () => ({
  useSelector: jest.fn().mockReturnValue(true),
  useDispatch: () => mockDispatch
}))

const component = <Modal fruit={fruit} />

describe('the modal component', () => {
  it('functions as expected', () => {
    const { getByTestId } = render(component)
    fireEvent.click(getByTestId('close'))

    expect(mockDispatch.mock.calls).toHaveLength(1)
  })

  it('renders the same as last time', () => {
    const { container } = render(component)
    expect(container).toMatchSnapshot()
  })
})
