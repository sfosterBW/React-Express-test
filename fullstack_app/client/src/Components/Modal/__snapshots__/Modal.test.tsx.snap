// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the modal component renders the same as last time 1`] = `
<div>
  <section
    className="wrapper"
  >
    <div
      className="modal"
    >
      <form
        className="form"
        onSubmit={[Function]}
      >
        <h2
          className="title"
        >
          Edit fruit
        </h2>
        <div
          className="wrapper"
        >
          <label
            className="label"
            htmlFor="name"
          >
            Add a fruit
          </label>
          <input
            className="text"
            id="name"
            name="name"
            onChange={[Function]}
            type="text"
            value="False Case"
          />
        </div>
        <div
          className="wrapper"
        >
          <label
            className="label"
            htmlFor="best"
          >
            Is it best?
          </label>
          <div
            className="checkboxWrapper"
          >
            <input
              checked={false}
              className="checkbox"
              id="best"
              name="best"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className="checkboxLabel"
              htmlFor="best"
              style={
                Object {
                  "background": undefined,
                }
              }
            >
              <span
                className="checkboxButton"
              />
            </label>
          </div>
        </div>
        <button
          className="button"
        >
          Edit fruit
        </button>
      </form>
      <button
        className="close"
        name="close"
        onClick={[Function]}
      >
        Ã—
      </button>
    </div>
  </section>
</div>
`;
